src/tools/general.js

```diff
@@ -147,71 +147,67 @@ function findLastReviewIndexByDate({ reviews, dateKey }) {
  */
-async function getReviews({ placeId, client }) {
-    /** @type {any[]} */
-    const result = [];
+async function getReviews({ placeId, client, session }) {
     let offset = 0;
     const limit = 20;
     const { maxReviews } = getConfig();
+    let reviewsCount = 0;
+    let reachedLimit = false;
 
-    while (true) {
-        const resp = await callForReview({ placeId, client, offset, limit });
-        const { errors } = resp ?? {};
+    const reviewsStore = await Apify.openKeyValueStore(`reviews-${placeId}`);
 
-        if (errors) {
-            log.error('Graphql error', { errors });
-        }
+    while (!reachedLimit) {
+        const resp = await callForReview({ placeId, session, client, offset, limit });
+        const { statusCode, body: { data, paging } } = resp ?? {};
 
-        if (!resp?.data?.locations?.length) {
-            throw new Error('Missing locations');
+        if (statusCode !== 200 || !data) {
+            throw new Error('Failed to get reviews');
         }
 
-        offset += limit;
-
-        const reviewData = resp?.data?.locations[0]?.reviewListPage || {};
+        let reviews = data;
+        const { total_results: totalCount, next } = paging;
 
-        const { totalCount } = reviewData;
-        let { reviews = [] } = reviewData;
         const lastIndexByDate = findLastReviewIndexByDate({ reviews, dateKey: 'publishedDate' });
-        const lastIndexByReviewsLimit = maxReviews > 0 ? maxReviews : -1;
-        const smallestIndex = getSmallestIndexGreaterThanEqualZero(lastIndexByDate, lastIndexByReviewsLimit);
-        const shouldSlice = smallestIndex >= 0;
-        if (shouldSlice) {
-            reviews = reviews.slice(0, smallestIndex);
+        if (lastIndexByDate >= 0) {
+            log.info('Getting the last review by date');
+            reviews = reviews.slice(0, lastIndexByDate);
+            reviewsCount -= (lastIndexByDate + 1);
         }
 
-        const numberOfReviews = (smallestIndex === -1 || totalCount < smallestIndex) ? totalCount : smallestIndex;
-
-        log.info(`Going to process ${numberOfReviews} reviews`);
-        log.debug('params', { smallestIndex, offset, totalCount });
-
-        if (reviews.length >= 1) {
-            reviews.forEach((review) => result.push(processReview(review)));
+        if (reviewsCount > maxReviews) {
+            log.info('Getting the last review by maxReviews limit');
+            const sliceIndex = reviewsCount - maxReviews;
+            reviews = reviews.splice(0, sliceIndex);
+            reviewsCount -= (sliceIndex + 1);
         }
 
-        if (reviews.length < limit || (maxReviews > 0 && result.length >= maxReviews) || shouldSlice) break;
-    }
+        reviewsCount += reviews.length;
+        log.info(`Processing ${reviewsCount} of ${totalCount} reviews for placeId ${placeId}`);
 
-    return result;
-}
+        offset += limit;
+        await Promise.all(reviews.map((review) => reviewsStore.setValue(`${review.id}`, review)));
```

Comment:
Please never store anything massively to KV store. You can hit a rate limit and it takes time. Also if you store it to named store, it means it will stay there forever.

Much better design is to have a `state` variable that you pass into functions and save everything there. Because JS is single threaded, it prevents any race conditions on a single variable. In `main.js` you just need to create that state and set a persistence like this:

```
// Load from store if the run was migrated or resurrected
const state = await Apify.getValue('STATE') || {};

Apify.on('persistState', async () => {
     // You can also log some info about the state here
     await Apify.setValue('STATE', state);
}
```